crypto/evp/evp_lib.c

```diff
@@ -114,6 +120,64 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
     return i;
 }
 
+int EVP_CIPHER_get_asn1_aead_params(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+{
+    int i = 0;
+    long tl;
+    unsigned char *iv;
+
+    if (type == NULL) {
+        return i;
+    }
+
+    if (!(c->cipher->flags & EVP_CIPH_FLAG_AEAD_CIPHER)) {
+        return -1;
+    }
+
+    i = ASN1_TYPE_get_octetstring_int(type, &tl, NULL, INT_MAX);
+    if (i <= 0 || (iv = OPENSSL_malloc(i)) == NULL)
+        return -1;
+    ASN1_TYPE_get_octetstring_int(type, &tl, iv, i);
+
+    if (EVP_CIPHER_CTX_ctrl(c, EVP_CTRL_AEAD_SET_IVLEN, i, NULL) <= 0
+            || EVP_CIPHER_CTX_ctrl(c, EVP_CTRL_SET_IV, i, iv) <= 0)
+        i = -1;
+
+    OPENSSL_free(iv);
+
+    return i;
+}
+
+
+int EVP_CIPHER_set_asn1_aead_params(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+{
+    int ivl, tl, ok = 0;
+    unsigned char *iv;
+
+    if (type == NULL) {
+        return 0;
+    }
+
+    if (!(c->cipher->flags & EVP_CIPH_FLAG_AEAD_CIPHER)) {
+        return -1;
+    }
+
+    tl = EVP_CIPHER_CTX_ctrl(c, EVP_CTRL_AEAD_GET_TAGLEN, 0, NULL);
+    if (tl <= 0)
+        return -1;
+
+    ivl = EVP_CIPHER_CTX_iv_length(c);
+    if (ivl <= 0 || (iv = OPENSSL_malloc(ivl)) == NULL)
```

Comment:
Similarly this could be stack allocated.

Reply:
Fixed